File: java/v1/src/main/java/com/kmeans/Kmeans.java
package com.kmeans;

import java.util.ArrayList;
import java.util.List;

public class Kmeans {
    
    private final int numClusters;
    private final List<Point> points;
    private List<List<Point>> clusters;
    private List<Point> centroids;
    
    public Kmeans(int numClusters, List<Point> points, long seed) {
        if(points == null || points.size() < numClusters){
            throw new IllegalArgumentException("Number of clusters greater than the number of points");
        }

        this.points = points;
        this.numClusters = numClusters;
        this.createClusters();
    }

    private void createClusters() {
        this.clusters = new ArrayList<>();
        for(int i = 0; i < this.numClusters; i++){
            this.clusters.add(new ArrayList<>());
        }
    }

    private void randCentroids() {
        this.centroids = new ArrayList<>();
        int[] indices = {10, 532, 9012};
        for (int i = 0; i < this.numClusters; i++) {
            this.centroids.add(new Point(this.points.get(indices[i])));;
        }
    }

    private void updateCentroids() {
        List<Point> newCentroids = new ArrayList<>();

        for (int i = 0; i < this.numClusters; i++) {
            List<Point> cluster = this.clusters.get(i);

            if (cluster.isEmpty()) {
                newCentroids.add(this.centroids.get(i));
                continue;
            }

            int dim = cluster.get(0).getDimension();
            Point sum = new Point(new double[dim]); 

            for (Point point : cluster) {
                sum.accumulatedAdd(point);
            }

            sum.accumulatedDivide(cluster.size());

            newCentroids.add(sum);
        }
        this.centroids = newCentroids;
    }

    private int findNearestCentroid(Point p) {
        double minDistance = Double.MAX_VALUE;
        int nearestIndex = -1;
        
        for (int i = 0; i < centroids.size(); i++) {
            double dist = p.sqDistanceTo(centroids.get(i));
            if (dist < minDistance) {
                minDistance = dist;
                nearestIndex = i;
            }
        }
        return nearestIndex;
    }

    private boolean centroidsConverged(List<Point> last, List<Point> now) {
        double epsilon = 1e-10;
        for (int i = 0; i < this.numClusters; i++) {
            if (last.get(i).sqDistanceTo(now.get(i)) > epsilon) {
                return false;
            }
        }
        return true;
    }

    public void fit() {
        this.fit(300);
    }

    public void fit(int maxIterations) {
        this.randCentroids();
        List<Point> lastCentroids;
        int iteration = 0;
        boolean converged = false;
        
        do {
            lastCentroids = new ArrayList<>(this.centroids);

            for (List<Point> cluster : this.clusters) {
                cluster.clear();
            }

            for (Point point : this.points) {
                int indexCluster = this.findNearestCentroid(point);
                this.clusters.get(indexCluster).add(point);
            }

            this.updateCentroids();
            converged = centroidsConverged(lastCentroids, this.centroids);
            iteration++;
        } while (!converged && iteration < maxIterations);
    }

    public List<Point> getCentroids() {
        return centroids;
    }

    public List<List<Point>> getClusters() {
        return clusters;
    }
}

File: go/v1/kmeans.go
package main

import (
	"errors"
	"math"
)

type kmeans struct {
	numClusters int
	points []Point
	clusters [][]Point
	centroids []Point
}

func Kmeans(numClusters int, points []Point, seed uint64) (*kmeans, error) {
	numPoints := len(points)
	if numClusters == 0 || numClusters > numPoints {
		return nil, errors.New("number of clusters greater than the number of points or zero points")
	}

	instance := &kmeans{
		numClusters: numClusters,
		points: points,
		clusters: make([][]Point, numClusters),
		centroids: make([]Point, numClusters),
	}

	indices := []int{10, 532, 9012}
	instance.initCentroids(indices)

	return instance, nil
}

func (k *kmeans) initCentroids(indices []int) {
	for i := 0; i < k.numClusters; i++ {
		k.centroids[i] = k.points[indices[i]]
	}
}

func (k *kmeans) updateCentroids() error {
	newCentroids := make([]Point, k.numClusters)

	for i := range k.numClusters {
		cluster := k.clusters[i]

		if len(cluster) == 0 {
			newCentroids[i] = k.centroids[i]
			continue
		}

		dim := len(cluster[0].position)
		sum, err := NewPoint(make([]float64, dim))

		if err != nil {
			return err
		}

		for _, point := range cluster {
			sum.AccumulatedAddPoint(&point)
		}

		sum.AccumulatedDivide(float64(len(cluster)))

		newCentroids[i] = *sum
	}

	copy(k.centroids, newCentroids)
	return nil
}

func (k *kmeans) findNearestCentroid(p *Point) (int, error) {
	minDistance := math.MaxFloat64
	nearestIndex := -1

	for i := range k.numClusters {
		dist, err := p.SqDistanceTo(&k.centroids[i])

		if err != nil {
			return -1, err
		}

		if dist < minDistance {
			minDistance = dist
			nearestIndex = i
		}
	}

	return nearestIndex, nil
}

func (k *kmeans) centroidsConverged(last []Point, now []Point) (bool, error) {
	epsilon := 1e-10
	for i := range k.numClusters {
		dist, err := last[i].SqDistanceTo(&now[i])

		if err != nil {
			return false, err
		}

		if  dist> epsilon {
			return false, nil
		}
	}
	return true, nil
}

func (k *kmeans) Fit() error {
	err := k.FitMaxIterations(300)
	
	if err != nil {
		return err
	}

	return nil
}

func (k *kmeans) FitMaxIterations(maxIterations int) error {
	iteration := 0
	lastCentroids := make([]Point, k.numClusters)

	for {
		copy(lastCentroids, k.centroids)

		for i := range k.clusters {
		    k.clusters[i] = k.clusters[i][:0]
		}

		for _, point := range k.points {
			indexCluster, err := k.findNearestCentroid(&point)

			if err != nil {
				return err
			}

			k.clusters[indexCluster] = append(k.clusters[indexCluster], point)
		}

		k.updateCentroids()
		converged, err := k.centroidsConverged(lastCentroids, k.centroids)

		if err != nil {
			return err
		}

		iteration++

		if converged || iteration >= maxIterations {
			break
		}
	}
	return nil
}

func (k *kmeans) GetCentroids() []Point {
	return k.centroids
}

func (k *kmeans) GetClusters() [][]Point {
	return k.clusters
}
